### PR TITLE
ICP-9908 Improve battery handling for IKEA Wireless Dimmer

### PR DESCRIPTION
**Pull request by dkirker**

### CODE REVIEW

File: devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: I'm not familiar with the values typically reported from this device but for other devices we've implemented an algorithm where the device needs to report a new value 2 (or more) times before it would be considered valid. That way if the value actually changes it does get reported eventually.

Comment (greens) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: Based on the issue it looks like these reports come too frequently to rely on that.

Comment (greens) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: what if we kept a running average of the last, say, ten values?

Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: A running average could work. It seems that the battery values fluctuate based on the usage at the time. I seem to recall that the device only responds to battery events when the device is being used.

Some of the reported values were: 37, 11, 37, 11, 37, 11, 17, 8, 11, 0, 11, 17, 11, 17.

I've never actually experienced reporting as frequent as certification has.

**Review by greens — DISMISSED** [review]
I guess an average of the most recent 2 values is probably the easiest way to do it. I suppose this will be more reactive if we start getting 0 readings.

File: devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy
Comment (dkirker) [inline devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: Was there a reason that the original DTH didn't use zigbee.batteryConfig? @MMateusiakS

Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: cc @tpmanley

**Review by greens — DISMISSED** [review]
This looks like a smart change.

**Review by tpmanley — DISMISSED** [review]

**Comment by dkirker** [comment]
This is still pointed at staging (I believe that was to get it in to staging after a deploy had occurred). If we would rather have this go to staging on July 16 instead of being merged and then deployed then, I can either rebase to master, or just wait until after the next deploy to merge.

**Comment by greens** [comment]
I'd feel better if it was pointed back at master.

**Comment by dkirker** [comment]
Done

File: devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: The report interval is longer than the checkInterval which will make the sengled switch get marked offline.

Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: I'm going to up the checkInterval, if that sounds good. I think 6 hours is plenty short enough, especially for battery devices.

Comment (tpmanley) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: Does the sengled respond to pings? If both the report interval and check interval are both 6 hours then one missed report could cause it to go offline if it doesn't respond to pings.

Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: I was going to set the checkInterval to 2 * reportInterval + ~2 mins to 10 mins. Unless you think reportInterval should be 3 hours and checkIntervale 6 hours + buffer? Or shorter?

Comment (tpmanley) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: I do think `checkInterval` should be 2 * reportInterval + 2-10 minutes. Typically we'd do 5 minutes for report interval but that impacts battery life so I think it's a good idea to increase it. I'd say go with a 3 hour reportInterval and 6:10 checkInterval for now and also run it by product for their thoughts.

Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: Made some updates. I realized the IKEA dimmer does report to us, we just can't ping it. So now the reportInterval and checkInterval are all in sync. I'm surprised that some of it wasn't caught before. Anyway, I have this running at home, so we'll see if it marks offline.

File: devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: sp: "health"

**Review by tpmanley — DISMISSED** [review]

File: devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy
Comment (dkirker) [reply devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy]: Fixed

**Review by tpmanley — APPROVED** [review]

**Pull request merged**
